### ISSUE TITLE
Starting interactive shell, without setting TERM, causes SIGSEGV

### ISSUE THREAD

**Issue by mojotx** [fish-shell/fish-shell#8873]
- Fish Version: `fish, version 3.4.1-168-g2bf9b97ab`
- OS: macOS

I recently noticed an [annoying issue](https://github.com/microsoft/vscode/issues/147320) when I tried launching the Microsoft VSCode editor from the macOS Dock. I was presented with an error, _"Unable to resolve your shell environment: Unexpected exit code from spawned shell (code null, signal SIGSEGV)."_

![Screen Shot 2022-04-12 at 13 57 12](https://user-images.githubusercontent.com/5694899/163263030-276ba306-7e6b-437a-a4a7-dbbf4b4d5589.png)

This was puzzling, because I could run a fish in a terminal session within the editor just fine. I did some digging through the Visual Studio Code codebase, and eventually figured out how to bump up the logging level, while still launching the editor as a macOS "app", since launching VSCode from the command-line would not trigger the bug. 

From vscode main.log:
``` 
[2022-04-13 14:08:12.112] [main] [trace] getUnixShellEnvironment#stderr warning: Could not set up terminal.
warning: TERM environment variable not set.
```

### TL;DR:
The bug is caused by the `TERM` environment variable being set. I then checked a Linux box running fish, to see if the problem was specific to macOS, and saw that the problem does NOT occur with the recent fish 3.4.1 release. 

I was then able to bisect the bug to commit 9c96986b3688974f9381e8509d66a73af1dc8a4f, but it's not immediately obvious why it's causing a SIGSEGV, since that was only a change to `src/builtins/set_color.cpp`.  Does `set_color` always get executed?

With fish prior to commit 9c96986b3688974f9381e8509d66a73af1dc8a4f:
```
$ sh -c 'env -i HOME=$(mktemp -d) /Users/mjarvis/src/github.com/fish-shell/build/fish'
warning: Could not set up terminal.
warning: TERM environment variable not set.
Welcome to fish, the friendly interactive shell
Type help for instructions on how to use fish
```

With fish >= fish 3.4.1-9c96986b3:
```
$ sh -c 'env -i HOME=$(mktemp -d) /Users/mjarvis/src/github.com/fish-shell/build/fish'
warning: Could not set up terminal.
warning: TERM environment variable not set.
fish: Job 1, 'sh -c 'env -i HOME=$(mktemp -d)…' terminated by signal SIGSEGV (Address boundary error)
```

**Comment by mojotx** [comment]
@floam -- Any ideas?

**Comment by morhook** [comment]
A very similar problem to this one is happening me on Ubuntu 22.04.01 (on Gnome). Having latest fish (3.5.1-1~jammy) and VSCode 1.71.0.
